### ISSUE TITLE
torch MLIR gotten from OPT ONNX model crashes IREE

### ISSUE THREAD

**Issue by kumardeepakamd** [nod-ai/iree-amd-aie#103]
### Steps to reproduce:
git clone https://github.com/nod-ai/SHARK-TestSuite.git
cd e2eshark
Run following command after providing your own values for :
-c : a valid torch MLIR build dir (build https://github.com/llvm/torch-mlir if you do not have a build already) : Note that torch MLIR is huge, hence should be generated.
-i : a valid IREE build dir (build https://github.com/openxla/iree f you do not have a build already)
--hfhome : A directory with more than 10GB of free space as hugging face models weights will be downloaded there

python ./run.py --upto inference --mode onnx -c ../../torch-mlir/build -i ../../iree-build/ --tests pytorch/models/opt-125M --hfhome /proj/gdba/kumar/HF_HOME

then cd to test-run/pytorch/models/opt-125M and you can find torch mlir and run logs. You can work in this directory to iterate on your fixes.

### Stack trace of crash of optimized build
iree-compile: iree/third_party/llvm-project/llvm/include/llvm/Support/Casting.h:566: decltype(auto) llvm::cast(const From &) [To = mlir::DenseElementsAttr, From = mlir::Attribute]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
Please report issues to https://github.com/openxla/iree/issues and include the crash backtrace.
Stack dump:
0.	Program arguments: /proj/gdba/kumar/nod/iree-build/tools/iree-compile --iree-hal-target-backends=llvm-cpu opt-125M.fp32.onnx.torch.mlir
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  libIREECompiler.so 0x00007fa14a046757 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  libIREECompiler.so 0x00007fa14a044970 llvm::sys::RunSignalHandlers() + 80
2  libIREECompiler.so 0x00007fa14a046e0f
3  libpthread.so.0    0x00007fa14413f420
4  libc.so.6          0x00007fa143ce900b gsignal + 203
5  libc.so.6          0x00007fa143cc8859 abort + 299
6  libc.so.6          0x00007fa143cc8729
7  libc.so.6          0x00007fa143cd9fd6
8  libIREECompiler.so 0x00007fa14ad4042e
9  libIREECompiler.so 0x00007fa14ad2d16b
10 libIREECompiler.so 0x00007fa14ad2cac9
11 libIREECompiler.so 0x00007fa14d3fa9e7
12 libIREECompiler.so 0x00007fa14d3f799e mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) + 942
13 libIREECompiler.so 0x00007fa14d3e3389
14 libIREECompiler.so 0x00007fa14d3dfa32 mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) + 1058
15 libIREECompiler.so 0x00007fa14d3a9ddb
16 libIREECompiler.so 0x00007fa14a1dc666 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 646
17 libIREECompiler.so 0x00007fa14a1dce78 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 344
18 libIREECompiler.so 0x00007fa14a1e23ae
19 libIREECompiler.so 0x00007fa14a1de3fb mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) + 2299
20 libIREECompiler.so 0x00007fa14a1dc801 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 1057
21 libIREECompiler.so 0x00007fa14a1dce78 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 344
22 libIREECompiler.so 0x00007fa14a1e16b1
23 libIREECompiler.so 0x00007fa14b1c1d64
24 libIREECompiler.so 0x00007fa14a1dc666 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 646
25 libIREECompiler.so 0x00007fa14a1dce78 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 344
26 libIREECompiler.so 0x00007fa14a1df269 mlir::PassManager::run(mlir::Operation*) + 985
27 libIREECompiler.so 0x00007fa149fa18b2 ireeCompilerInvocationPipeline + 3714
28 libIREECompiler.so 0x00007fa14a1a6be5
29 libIREECompiler.so 0x00007fa14a1a647a
30 libc.so.6          0x00007fa143cca083 __libc_start_main + 243
31 iree-compile       0x0000564142c0d72e
Aborted (core dumped)

**Comment by stellaraccident** [comment]
As an aside, of this is huge, it would be a good case to pipe clean getting external data working.

**Comment by MaheshRavishankar** [comment]
I am getting a different error. 

```
$ .../build/RelWithDebInfo/tools/iree-compile --iree-hal-target-backends=llvm-cpu -o output.vmfb stripped-opt-125M.fp32.onnx.torch.mlir
stripped-opt-125M.fp32.onnx.torch.mlir:326:12: error: failed to legalize operation 'torch.aten.cumsum' that was explicitly marked illegal
    %279 = torch.aten.cumsum %272, %278, %int4 : !torch.vtensor<[1,6],si64>, !torch.int, !torch.int -> !torch.vtensor<[1,6],si64>
           ^
stripped-opt-125M.fp32.onnx.torch.mlir:326:12: note: see current operation: %463 = "torch.aten.cumsum"(%181, %462, %179) : (!torch.vtensor<[1,6],si64>, !torch.int, !torch.int) -> !torch.vtensor<[1,6],si64>
```

This is the input model I am using

https://gist.githubusercontent.com/MaheshRavishankar/7ed3bf3314210a3e8e56da77e03982f7/raw/d21444fbd7cdc9eab4f96b0d99e0ea45fe4a2bd6/stripped-opt-125M.fp32.onnx.torch.mlir

**Comment by kumardeepakamd** [comment]
Okay, will get latest torch MLIR code and rebuild and retest later tonight.

**Comment by kumardeepakamd** [comment]
Strange. I got latest everything and it still crashes for me.

**Comment by MaheshRavishankar** [comment]
> Strange. I got latest everything and it still crashes for me.

You have a stale build somewhere is my guess

**Comment by kumardeepakamd** [comment]
if I try: torch-mlir-opt -convert-torch-onnx-to-torch **-convert-torch-to-linalg** opt-125M.fp32.torch-onnx.mlir i.e. try to lower to lilang using torch-mlir-opt, I get a different error than what Mahesh is getting:
opt-125M.fp32.torch-onnx.mlir:201:12: error: failed to legalize operation 'torch.aten.view' that was explicitly marked illegal
    %198 = torch.operator "onnx.Gather"(%0, %197) : (!torch.vtensor<[50272,768],f32>, !torch.vtensor<[1,6],si64>) -> !torch.vtensor<[1,6,768],f32>
           ^
opt-125M.fp32.torch-onnx.mlir:201:12: note: see current operation: %304 = "torch.aten.view"(%302, %303) : (!torch.vtensor<[6,768],f32>, !torch.list<int>) -> !torch.vtensor<[1,6,768],f32>

**Comment by renxida** [comment]
`Tip: pip install iree-compiler -f https://openxla.github.io/iree/pip-release-links.html into some temp env and <your temp python env>/iree-ir-tool strip-data opt-125M.fp32.onnx.torch.mlir -o stripped-opt-125M.fp32.onnx.torch.mlir will help you get smaller IR to deal with. thanks to Stella and Stan for this tip. `
-- Kumar on Discord

**Comment by renxida** [comment]
Getting same error as @kumardeepakamd 

Bash history:
```
python ./run.py --upto torch-mlir --mode onnx -c ../../torch-mlir/build  --tests pytorch/models/opt-125M --hfhome ~/hfhome
iree-ir-tool strip-data opt-125M.fp32.onnx.torch.mlir -o stripped-opt-125M.fp32.onnx.torch.mlir
~/torch-mlir/build/bin/torch-mlir-opt --convert-torch-to-linalg --convert-torch-to-tmtensor ./stripped-opt-125M.fp32.onnx.torch.mlir
```

**Comment by renxida** [comment]
Using `iree-compile --iree-hal-target-backends=llvm-cpu -o output.vmfb stripped-opt-125M.fp32.onnx.torch.mlir`, I get the same error as @MaheshRavishankar . Does `iree-compile` call torch mlir and `--convert-torch-to-linalg` and `--convert-torch-to-tmtensor` automatically? How can I check which version of `torch-mlir` is being used?

**Comment by newling** [comment]
`unimplemented: found unhandled case of expansion/collapse in aten.view` suggests an unhandled reshape. 6x768 -> 1x6x768 above should be fine though https://github.com/llvm/torch-mlir/pull/2829 any luck finding the culprit? Some debug build info for torch-mlir recently [here]( https://github.com/llvm/torch-mlir/blob/main/docs/development.md#flags-to-enable-mlir-debugging) @renxida

**Comment by newling** [comment]
We may be barking up the wrong tree looking at aten.view. 

With this command:
```
$TORCH_MLIR_BUILD_DIR/bin/torch-mlir-opt  --convert-torch-to-linalg opt-125M.fp32.onnx.torch.mlir
```

I hit the error

```
 error: failed to legalize operation 'torch.aten.view' that was explicitly marked illegal
    %223 = torch.aten.view %221, %222 : !torch.vtensor<[6,768],f32>, !torch.list<int> -> !torch.vtensor<[1,6,768],f32>
```
  
 But if I add canonicalization 
 
 ```
$TORCH_MLIR_BUILD_DIR/bin/torch-mlir-opt --canonicalize  --convert-torch-to-linalg  opt-125M.fp32.onnx.torch.mlir
```

Then I hit an assertion failure:

```
 torch-mlir-opt: /proj/gdba/jamesn/workspace/torch-mlir/externals/llvm-project/llvm/include/llvm/Support/Casting.h:566: decltype(auto) llvm::cast(const From &) [To = mlir::DenseElementsAttr, From = mlir::Attribute]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
```

We should be able to canonicalize the IR.

**Comment by renxida** [comment]
So far i've experienced 4 different kinds of errors
- Using mlir-opt direcltly with tolinalg and totmtensor gives me "view" op related errors (see https://gist.github.com/renxida/91b331231cd4a5cb008c8ec4ca39b086)
- Using iree-compile (older version) gives me "cumsum" related errors (same as Mahesh)
- Using iree-compile (build from source from main) gives me a crash during "mul" op (see https://gist.github.com/renxida/36b62ee1679403628fe5528ae0a04da8)
- Using SHARK-TestSuite `run.py --upto inference` gives same error as reported by (James Newling) when --canonicalize is supplied to torch-mlir-opt (see https://gist.github.com/renxida/4d2c76a6efac7028d0ed1f1a1b51769a)

**Comment by renxida** [comment]
Some commands to help reproduce
```
~/torch-mlir/build/bin/torch-mlir-opt --convert-torch-to-linalg --convert-torch-to-tmtensor --debug -mlir-disable-threading -mlir-print-ir-after-all ./stripped-opt-125M.fp32.onnx.torch.mlir &> /tmp/torchopt.out
~/iree-build/tools/iree-compile --iree-hal-target-backends=llvm-cpu -o output.vmfb stripped-opt-125M.fp32.onnx.torch.mlir --debug &> /tmp/failure-new.out
```

@Shukla-Gaurav

**Comment by renxida** [comment]
Looks like the issue is fixed by

https://github.com/llvm/torch-mlir/pull/2831
https://github.com/llvm/torch-mlir/pull/2830

Rob will be merging these soon. I will bump torch-mlir submodule today, and re-test.

🤞 hope this solved everything.

**Comment by renxida** [comment]
Merged Rob and Gaurav's patches. Now a problem with the loweirng of AtenCumsumOp to TMTensor is exposed. Triaging it here: https://github.com/llvm/torch-mlir/issues/2845

TLDR: ONNX forces the dim to be a variable, but TorchToTMTensor requires the dim to be a constant.
